iothub/device/src/Fx/Fx.cs

```diff
@@ -2,53 +2,48 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System;
-using System.Collections.ObjectModel;
+using System.Collections.Generic;
+using System.Linq;
 using System.Reflection;
 using System.Runtime.InteropServices;
 
 namespace Microsoft.Azure.Devices.Client
 {
     internal static class Fx
     {
+        private static readonly HashSet<Type> s_fatalExceptionTypes = new()
+        {
+            typeof(OperationCanceledException),
+            typeof(ObjectDisposedException),
+            typeof(ArgumentException),
+            typeof(InvalidOperationException),
+            typeof(InvalidOperationException),
+            typeof(OutOfMemoryException),
+            typeof(SEHException),
+            typeof(NullReferenceException),
+        };
+
         internal static bool IsFatal(Exception ex)
         {
-            while (ex != null)
+            if (s_fatalExceptionTypes.Any(fatalType => fatalType.IsAssignableFrom(ex.GetType())))
             {
-                if (ex is OutOfMemoryException
-                    || ex is SEHException
-                    || ex is NullReferenceException)
-                {
-                    return true;
-                }
-
-                // These exceptions aren't themselves fatal, but since the CLR uses them to wrap other exceptions,
-                // we want to check to see whether they've been used to wrap a fatal exception. If so, then they
-                // count as fatal.
-                if (ex is TypeInitializationException
-                    || ex is TargetInvocationException)
-                {
-                    ex = ex.InnerException;
```

Comment:
This was the only thing using while loop, so if instead we just use recursion as the other blocks do.